odoo/tools/populate.py

```diff
@@ -1,180 +1,360 @@
-    to a random integer between a and b included in each input dict.
-
-    :param int a: minimal random value
-    :param int b: maximal random value
-    :param int seed:
-    :returns: function of the form (iterator, field_name, model_name) -> values
-    :rtype: function (iterator, str, str) -> dict
-    """
-    def get_rand_int(random=None, **kwargs):
-        return random.randint(a, b)
-    return compute(get_rand_int, seed=seed)
-
-def randfloat(a, b, seed=None):
-    """ Return a factory for an iterator of values dicts that sets the field
-    to a random float between a and b included in each input dict.
-    """
-    def get_rand_float(random=None, **kwargs):
-        return random.uniform(a, b)
-    return compute(get_rand_float, seed=seed)
-
-def randdatetime(*, base_date=None, relative_before=None, relative_after=None, seed=None):
-    """ Return a factory for an iterator of values dicts that sets the field
-    to a random datetime between relative_before and relative_after, relatively to
-    base_date
-
-    :param datetime base_date: override the default base date if needed.
-    :param relativedelta|timedelta relative_after: range up which we can go after the
-         base date. If not set, defaults to 0, i.e. only in the past of reference.
-    :param relativedelta|timedelta relative_before: range up which we can go before the
-         base date. If not set, defaults to 0, i.e. only in the future of reference.
-    :param seed:
-    :return: iterator for random dates inside the defined range
-    """
-    base_date = base_date or datetime(2020, 1, 1)
-    seconds_before = relative_before and ((base_date + relative_before) - base_date).total_seconds() or 0
-    seconds_after = relative_after and ((base_date + relative_after) - base_date).total_seconds() or 0
-
-    def get_rand_datetime(random=None, **kwargs):
-        return base_date + relativedelta(seconds=random.randint(int(seconds_before), int(seconds_after)))
-    return compute(get_rand_datetime, seed=seed)
+_logger = logging.getLogger(__name__)
+
+# Min/Max value for a date/datetime field
+MIN_DATETIME = datetime((datetime.now() - relativedelta(years=4)).year, 1, 1)
+MAX_DATETIME = datetime.now()
+MIN_ROWS_PER_DAY = 1000
+
+
+def vary_date_field(env, model, factors, min_date=MIN_DATETIME, max_date=MAX_DATETIME):
+    """
+    Spreads uniformly dates in the date range specified by [min_date, max_date]
+    """
+    total_table_size = model.search_count([]) * factors[model]
+    if total_table_size <= MIN_ROWS_PER_DAY:
+        return SQL('now()::timestamp')
+    total_days = (max_date - min_date).days
+    rows_per_day = max(MIN_ROWS_PER_DAY, total_table_size // total_days + 1)
+    if total_table_size <= MIN_ROWS_PER_DAY * total_days:
+        min_date = min_date + relativedelta(days=(total_days - total_table_size // MIN_ROWS_PER_DAY))
+    return SQL("%s + (row_number() OVER() / %s) * interval '1 day'", min_date, rows_per_day)
```

Comment:
This logic is wrong:
- dates should depend on the current factor, not row_number() to ensure consistency between records from the same serie (stock.picking & stock.move)
- I would express dates with a relative number of days based on initial value of the field.
- MIN_ROWS_PER_DAY is weird: bellow 1.46M records it sets the date to today for all records: seems like an unnecessary heuristic

I would do something like this:
> date_field - interval 's // x days'

with:
- s the current serie from generate series
- x: factors[model] / (365*4)

No need for min_datetime, max_datetime, and min_rows_per_day

Reply:
There's an issue with the formula `date_field - interval '(factors - s)/ x days'`. 
When `factors[model] < (365*4)` there will be date gaps. For instance, if you populate `account.move` with a factor `2`, you will do `date_field - interval '730 days'` for `series_alias=1` and 
`date_field - interval '0 days'` for `series_alias=2`. This means records in 2022, no records in 2023 and records in 2024, which is a bit weird.

Taking `min(factors[model], 365*4)` fixes this gap issue.